api/provider/aws/job_runner.go

```
@@ -68,189 +69,196 @@ func (r *JobRunner) Run(j models.Job) error {
 
-	return r.jobStore.SetJobResult(jobID, task.TaskID)
+		return task.TaskID, nil, false
+	})
 }
 
-func (r *JobRunner) deleteDeploy(jobID, deployID string) error {
-	return r.deployProvider.Delete(deployID)
+func (r *JobRunner) deleteDeploy(jobID, deployID string) (string, error) {
+	return "", r.deployProvider.Delete(deployID)
 }
 
-func (r *JobRunner) deleteEnvironment(jobID, environmentID string) error {
+func (r *JobRunner) deleteEnvironment(jobID, environmentID string) (string, error) {
 	loadBalancers, err := r.loadBalancerProvider.List()
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	for _, loadBalancer := range loadBalancers {
 		if loadBalancer.EnvironmentID == environmentID {
-			if err := r.deleteLoadBalancer(jobID, loadBalancer.LoadBalancerID); err != nil {
-				return err
+			if _, err := r.deleteLoadBalancer(jobID, loadBalancer.LoadBalancerID); err != nil {
+				return "", err
 			}
 		}
 	}
 
 	services, err := r.serviceProvider.List()
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	for _, service := range services {
 		if service.EnvironmentID == environmentID {
-			if err := r.deleteService(jobID, service.ServiceID); err != nil {
-				return err
+			if _, err := r.deleteService(jobID, service.ServiceID); err != nil {
+				return "", err
 			}
 		}
 	}
 
 	tasks, err := r.taskProvider.List()
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	for _, task := range tasks {
 		if task.EnvironmentID == environmentID {
-			if err := r.deleteTask(jobID, task.TaskID); err != nil {
-				return err
+			if _, err := r.deleteTask(jobID, task.TaskID); err != nil {
+				return "", err
 			}
 		}
 	}
 
-	return catchAndRetry(15, func() (shouldRetry bool, err error) {
+	return catchAndRetry(time.Minute*15, func() (result string, err error, shouldRetry bool) {
+		log.Printf("[DEBUG] [JobRunner] Deleting environment %s", environmentID)
 		if err := r.environmentProvider.Delete(environmentID); err != nil {
-			switch err := err.(type) {
```

@sesh-kebab 
> Indicates leaky abstraction. Can this retry logic be encapsulated within the environment provider itself?

Ignoring error capturing, just doing "blind" retries to avoid abstraction leaking

---

I'd like to avoid that if possible. We had mixed retry logic before between the job runner and backend. I think keeping the retry logic isolated to one area of the code makes things much clearer.